src/js/express.js

```diff
@@ -0,0 +1,179 @@
+gpii.express.registerComponentLineage = function(childComponent, expressComponent) {
+    // Get the list of parents for this component
+    var segments = gpii.express.pathForComponent(childComponent);
+
+    var childNickname  = segments[segments.length - 1];
+
+    var expressSegments = gpii.express.pathForComponent(expressComponent);
+
+    // If we only have one segment more than express, we are a child of express itself.
+    // We need this to be able to handle cases in which express itself is a child component.
+    if (segments.length === expressSegments.length + 1) {
+        expressComponent.options.members.directChildrenOfInterest.push(childNickname);
+    }
+    // Otherwise, register the nickname of this component and its immediate parent path, each level will take care of the next highest.
+    else if (segments.length > expressSegments.length + 1) {
+        var parentPath = segments.slice(expressSegments.length, segments.length - 1).join(".");
+        if (expressComponent.options.members.childrenByParent[parentPath]) {
+            expressComponent.options.members.childrenByParent[parentPath].push(childNickname);
+        }
+        else {
+            expressComponent.options.members.childrenByParent[parentPath] = [childNickname];
+        }
+    }
+};
+
+// Wire a child to its immediate descendants.
+gpii.express.connectDirectDescendants = function(that, childComponent, childPath) {
+    var descendants =  that.options.members.childrenByParent[childPath];
+
+    // This component has descendants, wire them in first
+    if (descendants !== undefined) {
+        // Wire together one level of routers and middleware, then recurse
+        for (var a = 0; a < descendants.length; a++) {
+            var grandchildNickname  = descendants[a];
+            var grandchildComponent = childComponent[grandchildNickname];
+            if (fluid.hasGrade(grandchildComponent.options, "gpii.express.router")) {
+                // We have to wire our children in with our path to preserve relative pathing, so that their router will begin with our path.
+                childComponent.options.router.use(childComponent.options.path, grandchildComponent.options.router);
+
+                // Recurse from here on down.
+                that.connectDirectDescendants(grandchildComponent, childPath + "." + grandchildNickname);
+            }
+            else if (fluid.hasGrade(grandchildComponent.options, "gpii.express.middleware")) {
+                if (childComponent.options.router) {
+                    // We have to wire our children in with our path to preserve relative pathing, so that their router will begin with our path.
+                    childComponent.options.router.use(childComponent.options.path, grandchildComponent.getMiddlewareFunction());
+                }
+                else {
+                    throw(new Error({message: "A component must expose a router in order to work with child middleware components."}));
```

Comment:
We use fluid.fail in these cases - exception handling in JavaScript is a cursed technique which must never be used. As well as not working the way that developers in corporate languages (C++/Java/C#) expect, it is handled extremely badly by the optimisers in the JavaScript VMs.